File: Solution/src/com/solution/array/LocalMaxima.java
package com.solution.array;

import com.solution.main.*;

import java.util.*;

public class LocalMaxima {
    public void run() {
        Scanner sc = ReadInput.getScanner("localMaxima");
        int size = sc.nextInt();

        String array = sc.next();

        int[] arr = ReadInput.getIntArray(size, array);

        doMaxima(arr, 0, arr.length - 1);

    }

    private void doMaxima(int[] arr, int beginIndex, int endIndex) {

        int e_index = (beginIndex + endIndex) / 2;

        if (e_index > 0 && e_index < arr.length - 1) {

            if (arr[e_index] >= arr[e_index - 1] && arr[e_index] >= arr[e_index + 1]) {
                System.out.println(arr[e_index]);
                return;
            } else if (arr[e_index] > arr[e_index - 1] && arr[e_index] < arr[e_index + 1]) {
                doMaxima(arr, e_index + 1, endIndex);
            } else {
                doMaxima(arr, beginIndex, e_index - 1);
            }
        }
        //Edge case
        if (e_index == 0) {
            doMaxima(arr, e_index + 1, endIndex);
        }


    }
}

File: src/main/kotlin/com/cssolutions/array/LocalMaxima.kt
package com.cssolutions.array

import java.util.*
import com.cssolutions.main.ReadInput

class LocalMaxima {
    fun run() {
        val sc: Scanner = ReadInput.getScanner("localMaxima")
        val size: Int = sc.nextInt()
        val array: String = sc.next()
        val arr = ReadInput.getIntArray(size, array)
        doMaxima(arr, 0, arr.size - 1)
    }

    private fun doMaxima(arr: IntArray, beginIndex: Int, endIndex: Int) {
        val e_index = (beginIndex + endIndex) / 2
        if (e_index > 0 && e_index < arr.size - 1) {
            if (arr[e_index] >= arr[e_index - 1] && arr[e_index] >= arr[e_index + 1]) {
                System.out.println(arr[e_index])
                return
            } else if (arr[e_index] > arr[e_index - 1] && arr[e_index] < arr[e_index + 1]) {
                doMaxima(arr, e_index + 1, endIndex)
            } else {
                doMaxima(arr, beginIndex, e_index - 1)
            }
        }
        //Edge case
        if (e_index == 0) {
            doMaxima(arr, e_index + 1, endIndex)
        }
    }
}